File: 1002. Find Common Characters/commonChars.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
const getCharCount = (str) => {
  const ctarCnt = {};
  for (let i = 0; i < str.length; i++) {
    if (!ctarCnt[str[i]]) {
      ctarCnt[str[i]] = 0;
    }
    ctarCnt[str[i]]++;
  }

  return ctarCnt;
};

const commonChars = (words) => {
    const ctarCnt = getCharCount(words[0]);

    for (let i = 1; i < words.length; i++) {
      const wordCtarCnt = getCharCount(words[i]);
      for (let key in ctarCnt) {
        ctarCnt[key] = Math.min(ctarCnt[key], wordCtarCnt[key] || 0);
      }
    }

    const res = [];
    for (let key in ctarCnt) {
      for (let i = 0; i < ctarCnt[key]; i++) {
        res.push(key);
      }
    }

    return res;
};

const words = ["bella", "label", "roller"];
console.log('commonChars: ', commonChars(words));

File: 1002. Find Common Characters/commonChars.py
from collections import Counter
from typing import List

class Solution:
  def commonChars(self, words: List[str]) -> List[str]:
    char_cnt = Counter(words[0])
    
    for i in range(1, len(words)):
      current_word_cnt = Counter(words[i])
      for c in char_cnt:
        char_cnt[c] = min(char_cnt[c], current_word_cnt[c])
        
    res = []
    for c in char_cnt:
      for i in range(char_cnt[c]):
        res.append(c)
        
    return res
      

words = ["bella","label","roller"]
solution = Solution()
print('commonChars: ', solution.commonChars(words))
